### DISCUSSION TITLE
How to improve shell scripting DX?

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#3276] [General]
## Overview

In the upcoming *cloak* release (0.3) the primary use of Dagger will be to embed it in a script. In the case of shell scripts (bash, zsh, powershell etc) the simplest approach is to invoke the `dagger` CLI directly.

Therefore it is crucially important that the `dagger` CLI be very easy to script. In particular it should be very easy to make multiple calls to `dagger`, and connect their inputs and outputs in a pipeline. Also known as "stitching".

## Design options

There are several possible approaches, each with different tradeoffs.

### Option 1: inline graphql

This is the current de facto design:

* Embed graphql queries as strings in your shell script
* Invoke these queries with `cloak do`
* Query parameters can be passed with `--set`
* Raw response data is filtered through `jq` to retrieve the field you want.

Example courtesy of @kpenfound:

```bash
WORK=$(cloak do<<'EOF'
query {
    host {
        workdir {
            read {
                id
            }
        }
    }
}
EOF
)
WORKFS=$(echo -n $WORK | jq -r '.host.workdir.read.id')

HELLO=$(cloak --set "workfs=$WORKFS" do<<'EOF'
query ($workfs: FSID!) {
  core {
    image(ref: "alpine") {
      exec(input: {
        args: ["apk", "add", "make"]
      }) {
        fs {
            exec(input: {
            args: ["make", "hello"]
            workdir: "/src"
            mounts: [
                {
                    path: "/src"
                    fs: $workfs
                }
            ]
        }) {
            stdout
        }
        }
      }
    }
  }
}
EOF
)
OUT=$(echo -n $HELLO | jq -r '.core.image.exec.fs.exec.stdout')
echo $OUT
```

See the [original source code](https://github.com/kpenfound/dagger-demos/blob/main/make/hello/dagger.sh)

Some comments and feedback on this approach:

* From @samalba :
  > I like relying on jq for formatting the output. I prefer this than dagger trying to do any magic, jq is far more powerful and easy enough for simple selection.
  > The inlined queries does not have issues with escaping thanks to the --set cli arg

* From @marcosnils :
  > A single query can run multiple steps in parallel by leveraging graphql field aliases, it's important to document it so users can make the most while using it.
  > Outputs usually require third party tools to parse like jq, so it'd be great to help out here.
  > Writing queries in a bash script is error prone and clunky as 99% of the times it involves some heredoc + interpolation handling
  > Capturing errors and exit codes? i.e: exit code when a query fails vs could not connect to engine
  > We can't enforce that all users that write inlined queries will use the --set flag, specially for those coming with a long bash background since the "usual" way of performing similar things in bash is through heredocs and interpolations.

### Option 2: graphql document

In this option, you still write raw graphql queries but keep them in a separate graphql document, then reference them by name in the bash script.

Example (courtesy of @marcosnils):

```
-- queries.gql --
query Build {
  core {
    git(remote: "https://github.com/dagger/todoapp", ref: "cloak") {
      yarn(runArgs: "build") {
        id
      }
    }
  }
}

query Deploy($contents: String!, $token: SecretID!) {
  netlify {
    deploy(
      contents: $contents,
      token: $token,
      siteName: "jeremy-foo",
      subdir: "build",
    ) {
      deployURL
    }
  }
}
```

```bash
 RUN="cloak do -f ./queries.gql"

CONTENTS=$($RUN --filter '.core.git.yarn.id' Build)

URL=$($RUN --set "contents=$CONTENTS" --secret "token=$TOKEN" --filter '.netlify.deploy.deployURL' Deploy)
echo $URL
```

### Option 3: query builder

In this option, you don't write raw graphql queries. Instead, the queries are created dynamically by the `dagger` CLI from its arguments.

Example (imaginary):

```bash
dagger query \
  container from --url alpine \
  exec --args '["apk", "add", "make"]' \
  withMountedDirectory --path /src --source "$(dagger query host workdir)" \
  withWorkdir /src \
  exec --args '["make", "hello"]' \
  stdout
```

### Option 4: generate shell script

In this option, a bash SDK generates a shell script based on ... something. A graphql document perhaps?

```bash
#!/bin/bash

# This will define one bash function per query in the graph doc, with same names:
#   function makeHello()
#   function getWorkdir()
. $(dagger generate --sdk bash --source ./queries.gql)

makeHello --dir $(getWorkdir)
```

### Formatting of inputs & ouputs

> @marcosnils says:
> Outputs usually require third party tools to parse like jq, so it'd be great to help out here.

### Inline graphql queries considered harmful

> Writing queries in a bash script is error prone and clunky as 99% of the times it involves some heredoc + interpolation handling

Possible solutions:

1. Query file on the side
2. Smart schema introspection (a la hasura grapqurl)

### Session management

As @sipsma explained elsewhere:

> right now each CLI call is a different session, as opposed to engine.Start where everything happens in the same one

Follow-up questions:

* What are concrete consequences of each call being in a different session?
* Assuming we want to avoid those consequences, can we do so without bundling an entire shell? Perhaps optionally wrap your shell with an env-style wrapper? `dagger env -- bash` which would inject a `$DAGGER_SESSION` env variable?

**Comment by marcosnils** [comment]
Challenges I see particularly in Dagger vs scripting any other tools: 

- A single query can run multiple steps in parallel by leveraging graphql field aliases, it's important to document it so users can make the most while using it. 
- Outputs usually require third party tools to parse like `jq`, so it'd be great to help out here.
- Writing queries in a bash script is error prone and clunky as 99% of the times it involves some heredoc + interpolation handling
- Capturing errors and exit codes? i.e: exit code when a query fails vs could not connect to engine

**Comment by shykes** [comment]
> * A single query can run multiple steps in parallel by leveraging graphql field aliases, it's important to document it so users can make the most while using it.

I agree, but is there something we should do about this that is specific to shell scripting?

> * Outputs usually require third party tools to parse like `jq`, so it'd be great to help out here.

I think `dagger` should offer common helpers for input and output formats, since it's a fundamental part of stitching queries together. You should not require an external tool for very common uses of the CLI.

> * Writing queries in a bash script is error prone and clunky as 99% of the times it involves some heredoc + interpolation handling

You're thinking of proposing a queries file on the side? I think that's a viable approach and you should propose more details here :)

> * Capturing errors and exit codes? i.e: exit code when a query fails vs could not connect to engine

I guess we can copy whatever `curl` does here?

**Comment by mircubed** [comment]
@aluzzardi I'm worried about your bandwidth for this one? I will put it in stuck for now since I think this can be a blocker for release if you don't have someone to help you with it.

**Comment by samalba** [comment]
Please take a look at #3268. My experience so far:

- I like relying on `jq` for formatting the output. I prefer this than dagger trying to do any magic, `jq` is far more powerful and easy enough for simple selection.
- The inlined queries does not have issues with escaping thanks to the `--set` cli arg

There is room for improvement (like the session management for example) but it's totally usable with the current CLI and easy enough.

I would close this issue and instead open more scoped one later (for instance: implement the session management, write more guides and examples, etc...).

**Reply by shykes** [reply]
Good point, since there's no clear action item yet, I converted this to a discussion. We can use it to compare different options to improve the shell scripting DX. Maybe some more scoped issues will come out of it.

**Comment by shykes** [comment]
> * The inlined queries does not have issues with escaping thanks to the `--set` cli arg

@marcosnils does this address the issues you saw?

**Reply by marcosnils** [reply]
> @marcosnils does this address the issues you saw?

I wouldn't say it's an "issue" per-se but more like a UX / DX annoyance. We can't enforce that all users that write inlined queries will use the `--set` flag, specially for those coming with a long bash background since the "usual" way of performing similar things in bash is through heredocs and interpolations. 

My point was that by separating queries into a new file, you basically have no choice. The only way to supply arguments to queries it through the `--set` argument which makes the whole thing look way cleaner.

**Comment by sipsma** [comment]
Relevent issue: https://github.com/dagger/dagger/issues/3082

That was made a long time ago (in the previous cloak repo), but if nothing has changed then there's a problem with the CLI where it can't correctly parse lists of strings w/ `--set`. I.e. you can't really use alpine because it asks for a list of `pkgs`.

**Comment by samalba** [comment]
> That was made a long time ago (in the previous cloak repo), but if nothing has changed then there's a problem with the CLI where it can't correctly parse lists of strings w/ `--set`. I.e. you can't really use alpine because it asks for a list of `pkgs`.

True (although I did not need it), `--set` is only needed when you need to pass info in between queries (output of one query is needed as input for another one).

Given that we mostly need ids (secret, fs), the current support is enough short term. For Alpine, you can simply set your packages in the query itself ([example here](https://github.com/dagger/dagger/blob/6fbd88b0fedbf81e8e81ee13d6ff27c8547956b2/examples/bash/yarn-build.sh#L26)), you don't need `--set` for that, except if one day you dynamically generate your pkg list from another query, which would be kind of weird.

**Comment by marcosnils** [comment]
> ### Option 2: graphql document
> In this option, you still write raw graphql queries but keep them in a separate graphql document, then reference them by name in the bash script.
> 
> Example: _FIXME_.

example: 

```graphql
-- queries.gql --
query Build {
  core {
    git(remote: "https://github.com/dagger/todoapp", ref: "cloak") {
      yarn(runArgs: "build") {
        id
      }
    }
  }
}

query Deploy($contents: String!, $token: SecretID!) {
  netlify {
    deploy(
      contents: $contents,
      token: $token,
      siteName: "jeremy-foo",
      subdir: "build",
    ) {
      deployURL
    }
  }
}
```

```shell
RUN="cloak do -f ./queries.gql"

CONTENTS=$($RUN --filter '.core.git.yarn.id' Build)

URL=$($RUN --set "contents=$CONTENTS" --secret "token=$TOKEN" --filter '.netlify.deploy.deployURL' Deploy)
echo $URL
```

**Reply by shykes** [reply]
Thanks! I incorporated this.

**Comment by kpenfound** [comment]
Some really cool ideas! I'm going to try out converting the above demo to `Option 2` and see how it feels, but it seems like it would be a massive improvement.

@marcosnils can you tell me more about how "A single query can run multiple steps in parallel by leveraging graphql field aliases" means in relation to the bash example at the top? Like I could have the FSID in `fs: $workfs` actually alias to the query that resolves that FSID?

**Reply by shykes** [reply]
unfortunately not, you can’t reference one field of a query from within another field. With aliases you can basically “branch out” a query to get more outputs (like exec N commands in parallel). But you still need to stitch the usual way.

**Reply by kpenfound** [reply]
Here's a commit converting the demo to Option 2 https://github.com/kpenfound/dagger-demos/commit/5b34a5f5918dd15dd68123d5beb171ca074fdd74

Much better DX imo
